# Management - Fragmenter Validation 1.0

In [1]:
import qcportal as ptl
import pandas as pd
import datetime
import time

In [1]:
# connect without auth
# read only
client = ptl.FractalClient()

In [2]:
# connect with authentication, therefore write access
# don't use unless you plan to submit things
client = ptl.FractalClient.from_file()

In [3]:
client

FractalClient(server_name='The MolSSI QCArchive Server', address='https://api.qcarchive.molssi.org:443/', username='ddotson')

## Dataset tracking

In [6]:
ds = client.get_collection("TorsionDriveDataset", "OpenFF Fragmenter Validation 1.0")

In [7]:
ds.list_specifications()

Description
Name                                                                     
default                       Standard OpenFF torsiondrive specification.
B3LYP-d3bj/aug-cc-pvdz  OpenFF model B3LYP-d3bj/aug-cc-pvdz TorsionDri...
B3LYP-d3bj/aug-cc-pvtz  OpenFF model B3LYP-d3bj/aug-cc-pvtz TorsionDri...
B3LYP-d3bj/def2-tzvppd  OpenFF model B3LYP-d3bj/def2-tzvppd TorsionDri...
BLYP-d3bj/dzvp          OpenFF model BLYP-d3bj/dzvp TorsionDrive explo...
BLYP-d3bj/aug-cc-pvdz   OpenFF model BLYP-d3bj/aug-cc-pvdz TorsionDriv...
BLYP-d3bj/aug-cc-pvtz   OpenFF model BLYP-d3bj/aug-cc-pvtz TorsionDriv...
BLYP-d3bj/def2-tzvppd   OpenFF model BLYP-d3bj/def2-tzvppd TorsionDriv...
LRC-WPBEH/dzvp          OpenFF model LRC-WPBEH/dzvp TorsionDrive explo...
LRC-WPBEH/aug-cc-pvdz   OpenFF model LRC-WPBEH/aug-cc-pvdz TorsionDriv...
LRC-WPBEH/aug-cc-pvtz   OpenFF model LRC-WPBEH/aug-cc-pvtz TorsionDriv...
LRC-WPBEH/def2-tzvppd   OpenFF model LRC-WPBEH/def2-tzvppd TorsionDriv...
MN15-d3bJ/dzvp          OpenFF model MN15-d3bJ/dzvp TorsionDrive explo...
MN15-d3bJ/aug-cc-pvdz   OpenFF model MN15-d3bJ/aug-cc-pvdz TorsionDriv...
MN15-d3bJ/aug-cc-pvtz   OpenFF model MN15-d3bJ/aug-cc-pvtz TorsionDriv...
MN15-d3bJ/def2-tzvppd   OpenFF model MN15-d3bJ/def2-tzvppd TorsionDriv...
TPSSH-d3bJ/dzvp         OpenFF model TPSSH-d3bJ/dzvp TorsionDrive expl...
TPSSH-d3bJ/aug-cc-pvdz  OpenFF model TPSSH-d3bJ/aug-cc-pvdz TorsionDri...
TPSSH-d3bJ/aug-cc-pvtz  OpenFF model TPSSH-d3bJ/aug-cc-pvtz TorsionDri...
TPSSH-d3bJ/def2-tzvppd  OpenFF model TPSSH-d3bJ/def2-tzvppd TorsionDri...

'default' always means 'b3lyp-d3bj' for us.

In [12]:
ds.status("default")

default
COMPLETE       16
ERROR           3
RUNNING         3

In [13]:
ds.df

default
Cc1c[c:2]([c:1]2ccccc2c1OP(=O)([O-])[O-])[O:3][...  TorsionDriveRecord(id='19712153', status='COMP...
c1cc[c:1]2c(c1)ccc[c:2]2[O:3][P:4](=O)([O-])[O-]    TorsionDriveRecord(id='20548975', status='RUNN...
c1ccc(cc1)C[CH2:1][NH:2]/[C:3](=[NH+:4]/C(=[NH2...  TorsionDriveRecord(id='19712155', status='COMP...
C[CH2:1][NH:2]/[C:3](=[NH+:4]/C=[NH2+])/N           TorsionDriveRecord(id='20548976', status='COMP...
C[CH2:1][NH:2]/[C:3](=[NH+:4]/C)/N                  TorsionDriveRecord(id='20548977', status='COMP...
Cc1c([nH]cn1)CSC[CH2:1][NH:2]/[C:3](=N\C#N)/[NH...  TorsionDriveRecord(id='19712158', status='COMP...
C[CH2:1][NH:2]/[C:3](=[N:4]\C#N)/NC                 TorsionDriveRecord(id='20548978', status='COMP...
C[CH2:1][NH:2]/[C:3](=[N:4]\C)/NC                   TorsionDriveRecord(id='20548979', status='COMP...
c1cc(oc1)[CH2:1][NH:2][c:3]2cc(c(c[c:4]2C(=O)[O...  TorsionDriveRecord(id='19712161', status='COMP...
c1cc(oc1)[CH2:4][NH:3][c:2]2cc(c(c[c:1]2C(=O)[O...  TorsionDriveRecord(id='19712162', status='COMP...
c1cc[c:2]([c:1](c1)C(=O)[O-])[NH:3][CH2:4]c2ccco2   TorsionDriveRecord(id='20548992', status='COMP...
c1cc(oc1)CNc2c[c:1]([c:2](cc2C(=O)[O-])[S:3](=[...  TorsionDriveRecord(id='19712164', status='COMP...
c1c[c:1]([c:2](cc1C(=O)[O-])[S:3](=[O:4])(=O)N)Cl   TorsionDriveRecord(id='20548993', status='COMP...
c1cc[c:1]([c:2](c1)[S:3](=[O:4])(=O)N)Cl            TorsionDriveRecord(id='20548994', status='ERROR')
CC[NH+](CC)C[CH2:4][O:3][C:2](=[O:1])c1ccc(cc1Cl)N  TorsionDriveRecord(id='19712167', status='COMP...
c1ccc(c(c1)[C:3](=[O:4])[O:2][CH2:1]C[NH3+])Cl      TorsionDriveRecord(id='20548995', status='COMP...
C[CH2:1][O:2][C:3](=[O:4])c1ccccc1Cl                TorsionDriveRecord(id='20548996', status='COMP...
c1c(cc([c:2]([c:1]1I)[O:3][c:4]2cc(c(c(c2)I)[O-...  TorsionDriveRecord(id='19712170', status='RUNN...
Cc1c[c:1]([c:2](c(c1)I)[O:3][c:4]2cc(c(c(c2)I)[...  TorsionDriveRecord(id='20548997', status='ERROR')
c1cc[c:1](cc1)[O:2][c:3]2c(ccc[c:4]2I)I             TorsionDriveRecord(id='20548998', status='ERROR')
CC(=O)Oc1ccccc1[C:1](=O)[NH:2][c:3]2ncc([s:4]2)...  TorsionDriveRecord(id='19712173', status='RUNN...
CC(=O)Oc1ccccc1[C:1](=O)[NH:2][c:3]2ncc[s:4]2       TorsionDriveRecord(id='20548999', status='COMP...

## Problem cases


From 'OpenFF Fragmenter Validation 1.0', job index: c1ccc:1Cl job id: '20548994'. Error message:
```
ComputeError(error_type='CancelledError', error_message='Caught Executor Error:\nTraceback (most recent call last):\n  File "/home/chodera/miniconda/envs/qcfractal/lib/python3.6/site-packages/qcfractal/queue/executor_adapter.py", line 15, in _get_future\n    return future.result()\n  File "/home/chodera/miniconda/envs/qcfractal/lib/python3.6/concurrent/futures/_base.py", line 423, in result\n    raise CancelledError()\nconcurrent.futures._base.CancelledError\n')
)
```

In [36]:
erred = [i for i in ds.df.default.values if i.status == 'ERROR']
erred

[TorsionDriveRecord(id='20548994', status='ERROR'),
 TorsionDriveRecord(id='20548997', status='ERROR'),
 TorsionDriveRecord(id='20548998', status='ERROR')]

Start with the first one:

In [37]:
tdr = erred[0]

In [49]:
tdr.id

'20548994'

In [50]:
client.modify_services(operation='restart', procedure_id=tdr.id)

Data(n_updated=1)

In [51]:
ds = client.get_collection("TorsionDriveDataset", "OpenFF Fragmenter Validation 1.0")

In [52]:
ds.list_specifications()

Description
Name                                                                     
default                       Standard OpenFF torsiondrive specification.
B3LYP-d3bj/aug-cc-pvdz  OpenFF model B3LYP-d3bj/aug-cc-pvdz TorsionDri...
B3LYP-d3bj/aug-cc-pvtz  OpenFF model B3LYP-d3bj/aug-cc-pvtz TorsionDri...
B3LYP-d3bj/def2-tzvppd  OpenFF model B3LYP-d3bj/def2-tzvppd TorsionDri...
BLYP-d3bj/dzvp          OpenFF model BLYP-d3bj/dzvp TorsionDrive explo...
BLYP-d3bj/aug-cc-pvdz   OpenFF model BLYP-d3bj/aug-cc-pvdz TorsionDriv...
BLYP-d3bj/aug-cc-pvtz   OpenFF model BLYP-d3bj/aug-cc-pvtz TorsionDriv...
BLYP-d3bj/def2-tzvppd   OpenFF model BLYP-d3bj/def2-tzvppd TorsionDriv...
LRC-WPBEH/dzvp          OpenFF model LRC-WPBEH/dzvp TorsionDrive explo...
LRC-WPBEH/aug-cc-pvdz   OpenFF model LRC-WPBEH/aug-cc-pvdz TorsionDriv...
LRC-WPBEH/aug-cc-pvtz   OpenFF model LRC-WPBEH/aug-cc-pvtz TorsionDriv...
LRC-WPBEH/def2-tzvppd   OpenFF model LRC-WPBEH/def2-tzvppd TorsionDriv...
MN15-d3bJ/dzvp          OpenFF model MN15-d3bJ/dzvp TorsionDrive explo...
MN15-d3bJ/aug-cc-pvdz   OpenFF model MN15-d3bJ/aug-cc-pvdz TorsionDriv...
MN15-d3bJ/aug-cc-pvtz   OpenFF model MN15-d3bJ/aug-cc-pvtz TorsionDriv...
MN15-d3bJ/def2-tzvppd   OpenFF model MN15-d3bJ/def2-tzvppd TorsionDriv...
TPSSH-d3bJ/dzvp         OpenFF model TPSSH-d3bJ/dzvp TorsionDrive expl...
TPSSH-d3bJ/aug-cc-pvdz  OpenFF model TPSSH-d3bJ/aug-cc-pvdz TorsionDri...
TPSSH-d3bJ/aug-cc-pvtz  OpenFF model TPSSH-d3bJ/aug-cc-pvtz TorsionDri...
TPSSH-d3bJ/def2-tzvppd  OpenFF model TPSSH-d3bJ/def2-tzvppd TorsionDri...

'default' always means 'b3lyp-d3bj' for us.

In [54]:
ds.status("default")

default
COMPLETE       16
RUNNING         4
ERROR           2

In [56]:
ds.df

default
Cc1c[c:2]([c:1]2ccccc2c1OP(=O)([O-])[O-])[O:3][...  TorsionDriveRecord(id='19712153', status='COMP...
c1cc[c:1]2c(c1)ccc[c:2]2[O:3][P:4](=O)([O-])[O-]    TorsionDriveRecord(id='20548975', status='RUNN...
c1ccc(cc1)C[CH2:1][NH:2]/[C:3](=[NH+:4]/C(=[NH2...  TorsionDriveRecord(id='19712155', status='COMP...
C[CH2:1][NH:2]/[C:3](=[NH+:4]/C=[NH2+])/N           TorsionDriveRecord(id='20548976', status='COMP...
C[CH2:1][NH:2]/[C:3](=[NH+:4]/C)/N                  TorsionDriveRecord(id='20548977', status='COMP...
Cc1c([nH]cn1)CSC[CH2:1][NH:2]/[C:3](=N\C#N)/[NH...  TorsionDriveRecord(id='19712158', status='COMP...
C[CH2:1][NH:2]/[C:3](=[N:4]\C#N)/NC                 TorsionDriveRecord(id='20548978', status='COMP...
C[CH2:1][NH:2]/[C:3](=[N:4]\C)/NC                   TorsionDriveRecord(id='20548979', status='COMP...
c1cc(oc1)[CH2:1][NH:2][c:3]2cc(c(c[c:4]2C(=O)[O...  TorsionDriveRecord(id='19712161', status='COMP...
c1cc(oc1)[CH2:4][NH:3][c:2]2cc(c(c[c:1]2C(=O)[O...  TorsionDriveRecord(id='19712162', status='COMP...
c1cc[c:2]([c:1](c1)C(=O)[O-])[NH:3][CH2:4]c2ccco2   TorsionDriveRecord(id='20548992', status='COMP...
c1cc(oc1)CNc2c[c:1]([c:2](cc2C(=O)[O-])[S:3](=[...  TorsionDriveRecord(id='19712164', status='COMP...
c1c[c:1]([c:2](cc1C(=O)[O-])[S:3](=[O:4])(=O)N)Cl   TorsionDriveRecord(id='20548993', status='COMP...
c1cc[c:1]([c:2](c1)[S:3](=[O:4])(=O)N)Cl            TorsionDriveRecord(id='20548994', status='RUNN...
CC[NH+](CC)C[CH2:4][O:3][C:2](=[O:1])c1ccc(cc1Cl)N  TorsionDriveRecord(id='19712167', status='COMP...
c1ccc(c(c1)[C:3](=[O:4])[O:2][CH2:1]C[NH3+])Cl      TorsionDriveRecord(id='20548995', status='COMP...
C[CH2:1][O:2][C:3](=[O:4])c1ccccc1Cl                TorsionDriveRecord(id='20548996', status='COMP...
c1c(cc([c:2]([c:1]1I)[O:3][c:4]2cc(c(c(c2)I)[O-...  TorsionDriveRecord(id='19712170', status='RUNN...
Cc1c[c:1]([c:2](c(c1)I)[O:3][c:4]2cc(c(c(c2)I)[...  TorsionDriveRecord(id='20548997', status='ERROR')
c1cc[c:1](cc1)[O:2][c:3]2c(ccc[c:4]2I)I             TorsionDriveRecord(id='20548998', status='ERROR')
CC(=O)Oc1ccccc1[C:1](=O)[NH:2][c:3]2ncc([s:4]2)...  TorsionDriveRecord(id='19712173', status='RUNN...
CC(=O)Oc1ccccc1[C:1](=O)[NH:2][c:3]2ncc[s:4]2       TorsionDriveRecord(id='20548999', status='COMP...